drivers/platform/stm32/stm32_usb_uart.c

```diff
@@ -0,0 +1,252 @@
+{
+	USBD_CDC_SetTxBuffer(&hUsbDeviceFS, UserTxBufferFS, 0);
+	USBD_CDC_SetRxBuffer(&hUsbDeviceFS, UserRxBufferFS);
+
+	return USBD_OK;
+}
+
+static int8_t CDC_DeInit_FS(void)
+{
+	return USBD_OK;
+}
+
+static int8_t CDC_Control_FS(uint8_t cmd, uint8_t* pbuf, uint16_t length)
+{
+	return USBD_OK;
+}
+
+void stm32_on_usb_cdc_acm_rx(uint8_t* buf, uint32_t len)
+{
+	unsigned int i = 0;
+	int ret;
+
+	while(i < len) {
+		ret = lf256fifo_write(gfifo, buf[i]);
+		if (ret)
+			break;
+
+		i++;
+	}
+
+	return i;
+}
+
+static int8_t CDC_Receive_FS(uint8_t* Buf, uint32_t *Len)
+{
+	stm32_on_usb_cdc_acm_rx(Buf, *Len);
+
+	USBD_CDC_SetRxBuffer(&hUsbDeviceFS, &Buf[0]);
+	USBD_CDC_ReceivePacket(&hUsbDeviceFS);
+
+	return USBD_OK;
+}
+
+static uint8_t CDC_Transmit_FS(uint8_t* Buf, uint16_t Len)
+{
+	uint8_t result = USBD_OK;
+	USBD_CDC_HandleTypeDef *hcdc = (USBD_CDC_HandleTypeDef*)hUsbDeviceFS.pClassData;
+	if (hcdc->TxState != 0)
+		return USBD_BUSY;
+
+	USBD_CDC_SetTxBuffer(&hUsbDeviceFS, Buf, Len);
+	result = USBD_CDC_TransmitPacket(&hUsbDeviceFS);
+
+	return result;
+}
+
+static int8_t CDC_TransmitCplt_FS(uint8_t *Buf, uint32_t *Len, uint8_t epnum)
+{
+	tx_pending = 0;
+
+	return USBD_OK;
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc - The UART descriptor.
+ * @param param - The structure that contains the UART parameters.
+ * @return 0 in case of success, error code otherwise.
+ */
+static int32_t stm32_usb_uart_init(struct no_os_uart_desc **desc,
+				   struct no_os_uart_init_param *param)
+{
+	int ret;
+	struct no_os_uart_desc *descriptor;
+	struct stm32_usb_uart_desc *sdesc;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	sdesc = no_os_calloc(1, sizeof(*sdesc));
+	if (!sdesc) {
+		ret = -ENOMEM;
+		goto err_sdesc;
+	}
+	descriptor->extra = sdesc;
+
+	ret = lf256fifo_init(&sdesc->fifo);
+	if (ret)
+		goto err_fifo;
+
+	gfifo = sdesc->fifo;
+
+	*desc = descriptor;
+
+	return 0;
+err_fifo:
+	no_os_free(sdesc);
+err_sdesc:
+	no_os_free(descriptor);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by stm32_uart_init().
+ * @param desc - The UART descriptor.
+ * @return 0 in case of success, -1 otherwise.
+ */
+static int32_t stm32_usb_uart_remove(struct no_os_uart_desc *desc)
+{
+	struct stm32_usb_uart_desc *sdesc = desc->extra;
+	lf256fifo_remove(sdesc->fifo);
+	no_os_free(desc->extra);
+	no_os_free(desc);
+
+	return 0;
+};
+
+/**
+ * @brief Write data to UART device.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to write.
+ * @return Positive number of transmitted bytes in case of success, negative error code otherwise.
+ */
+static int32_t stm32_usb_uart_write(struct no_os_uart_desc *desc,
+				    const uint8_t *data,
+				    uint32_t bytes_number)
+{
+	unsigned int len = no_os_min(bytes_number, 2048);
+	while(tx_pending)
+		;
+	memcpy(UserTxBufferFS, data, len);
```

Comment:
Since the user provided `data` is const, I think we can avoid the memcpy() and use that buffer directly.

Reply:
The transmit part uses DMA and is nonblocking so I use a combination of a memcpy into separate buffer and the tx_complete flag to make sure the buffer is available for the DMA until TX is complete. In other words, if I don't use the memcpy, behavior is not ok, I see mixed up content if there are subsequent calls to the Tx function.

This is the only comment from your review that I won't implement, the rest I'm implementing right now.